charon/src/remove_read_discriminant.rs

```diff
@@ -48,67 +48,74 @@ impl<'a, 'tcx, 'ctx> Visitor<'a, 'tcx, 'ctx> {
-                                    variants
-                                        .iter()
-                                        .enumerate()
-                                        .all(|(i, _)| covered_variants.contains(&i))
-                                }
+                // Lookup the type of the scrutinee
+                let variants = match self.ctx.type_decls.get(*adt_id) {
+                    // This can happen if there was an error while extracting the definitions
+                    None => None,
+                    Some(d) => {
+                        match &d.kind {
+                            TypeDeclKind::Struct(_) | TypeDeclKind::Opaque => {
+                                // We shouldn't get there
+                                register_error_or_panic!(
+                                    self.ctx,
+                                    st.meta.span.rust_span_data.span(),
+                                    "Unreachable case"
+                                );
+                                None
                             }
+                            TypeDeclKind::Error(_) => None,
+                            TypeDeclKind::Enum(variants) => Some(variants),
                         }
                     }
                 };
+                let Some(variants) = variants else {
+                    // An error occurred. We can't keep the `Rvalue::Discriminant` around so we
+                    // `Nop` the whole statement sequence.
+                    assert!(self.ctx.error_count > 0);
+                    *st = Statement {
+                        content: RawStatement::Nop,
+                        meta: st.meta,
+                    };
+                    return
+                };
+
+                // Convert between discriminants and variant indices. Remark: the discriminant can
+                // be of any *signed* integer type (`isize`, `i8`, etc.).
+                let discr_to_id: HashMap<u128, VariantId::Id> = variants
+                    .iter_indexed_values()
+                    .map(|(id, variant)| (variant.discriminant, id))
+                    .collect();
+                let mut covered_discriminants: HashSet<u128> = HashSet::default();
+                let targets = targets
+                    .into_iter()
+                    .map(|(v, e)| {
+                        (
+                            v.into_iter()
+                                .map(|x| {
+                                    let discr = x.to_bits();
+                                    covered_discriminants.insert(discr);
+                                    discr_to_id.get(&discr).unwrap_or_else(|| {
+                                        panic!(
+                                            "Found incorrect discriminant {discr} for enum {}",
+                                            adt_id
+                                        )
```

Comment:
Can you use a register_error_or_panic so that we can recover from the error and print a message to the user?